File: cpp/RoundTrip/src/pong.cpp
/*                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2018 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

/**
 * This is a simple throughput application measuring obtainable throughput using the thingSDK
 *
 */
#include <iostream>
#include <thread>
#include <chrono>

#include <IoTDataThing.hpp>
#include <JSonThingAPI.hpp>
#include <ThingAPIException.hpp>
#include "include/cxxopts.h"
#include "include/utilities.h"

using namespace std;
using namespace com::adlinktech::datariver;
using namespace com::adlinktech::iot;

#ifdef _WIN32
static bool ctrlHandler(DWORD fdwCtrlType)
{
    cout << "Ctrl-c detected .. exiting" << endl;
	ExitProcess(1);
    return true;
}
#else
static void ctrlHandler(int fdwCtrlType)
{
    cout << "Ctrl-c detected .. exiting" << endl;
    exit(1);
}
#endif

class Pong {
private:
    string m_thingPropertiesUri;
    DataRiver m_dataRiver = createDataRiver();
    Thing m_thing = createThing();

    DataRiver createDataRiver() {
        return DataRiver::getInstance();
    }

    Thing createThing() {
        // Create and Populate the TagGroup registry with JSON resource files.
        JSonTagGroupRegistry tgr;
        tgr.registerTagGroupsFromURI("file://definitions/TagGroup/com.adlinktech.example/PingTagGroup.json");
        tgr.registerTagGroupsFromURI("file://definitions/TagGroup/com.adlinktech.example/PongTagGroup.json");
        m_dataRiver.addTagGroupRegistry(tgr);

        // Create and Populate the ThingClass registry with JSON resource files.
        JSonThingClassRegistry tcr;
        tcr.registerThingClassesFromURI("file://definitions/ThingClass/com.adlinktech.example/PongThingClass.json");
        m_dataRiver.addThingClassRegistry(tcr);

        // Create a Thing based on properties specified in a JSON resource file.
        JSonThingProperties tp;
        tp.readPropertiesFromURI(m_thingPropertiesUri);

        return m_dataRiver.createThing(tp);
    }

public:
    Pong(string thingPropertiesUri) :
        m_thingPropertiesUri(thingPropertiesUri)
    {
        cout << "Pong started" << endl;
    }

    ~Pong() {
        m_dataRiver.close();
        cout << "Pong stopped" << endl;
    }

    int run() {
        bool terminate = false;
        cout << "Waiting for samples from ping to send back..." << endl;

        while (!terminate) {
        	vector<DataSample<IOT_NVP_SEQ> > samples = m_thing.read<IOT_NVP_SEQ>("Ping");

            for (const DataSample<IOT_NVP_SEQ>& sample : samples) {
            	if (sample.getFlowState() == FlowState::PURGED) {
                    cout << "Received termination request. Terminating." << endl;
                    terminate = true;
                    break;
                } else {
                	m_thing.write("Pong", sample.getData());
                }
            }
        }

        return 0;
    }
};


int main(int argc, char *argv[]) {
    // Register handler for Ctrl-C
    registerControlHandler();

    try
    {
        return Pong("file://./config/PongProperties.json").run();
    }
    catch (ThingAPIException e)
    {
        cerr << "An unexpected error occurred: " << e.what() << endl;
    }catch(std::exception& e1){
        cerr << "An unexpected error occurred: " << e1.what() << endl;
    }

    unregisterControlHandler();
}

File: python/RoundTrip/pong.py
#!/usr/bin/env python

#
#                           ADLINK Edge SDK
#  
#     This software and documentation are Copyright 2018 to 2020 ADLINK
#     Technology Limited, its affiliated companies and licensors. All rights
#     reserved.
#  
#     Licensed under the Apache License, Version 2.0 (the "License");
#     you may not use this file except in compliance with the License.
#     You may obtain a copy of the License at
#  
#         http://www.apache.org/licenses/LICENSE-2.0
#  
#     Unless required by applicable law or agreed to in writing, software
#     distributed under the License is distributed on an "AS IS" BASIS,
#     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#     See the License for the specific language governing permissions and
#     limitations under the License.
#

'''
This is a simple throughput application measuring obtainable throughput using the thingSDK
'''

from __future__ import print_function
import sys
from adlinktech.datariver import DataRiver, JSonTagGroupRegistry, JSonThingClassRegistry, JSonThingProperties, FlowState

 
class Pong(object):
     
    # Initializing
    def __init__(self, thing_properties_uri):
        self._thing_properties_uri = thing_properties_uri
        
        self._datariver = None
        self._thing = None

    # Enter the runtime context related to the object
    def __enter__(self):
        self._datariver = DataRiver.get_instance()
        self._thing = self.create_thing()
        
        print('Pong started')
        
        return self

    # Exit the runtime context related to the object
    def __exit__(self, exc_type, exc_value, exc_traceback):
        if self._datariver is not None:
            self._datariver.close()
        print('Pong stopped')

    def create_thing(self):
        # Create and Populate the TagGroup registry with JSON resource files.
        tgr = JSonTagGroupRegistry()
        tgr.register_tag_groups_from_uri('file://definitions/TagGroup/com.adlinktech.example/PingTagGroup.json')
        tgr.register_tag_groups_from_uri('file://definitions/TagGroup/com.adlinktech.example/PongTagGroup.json')
        self._datariver.add_tag_group_registry(tgr) 
  
        # Create and Populate the ThingClass registry with JSON resource files.
        tcr = JSonThingClassRegistry()
        tcr.register_thing_classes_from_uri('file://definitions/ThingClass/com.adlinktech.example/PongThingClass.json')
        self._datariver.add_thing_class_registry(tcr)
  
        # Create a Thing based on properties specified in a JSON resource file.
        tp = JSonThingProperties()
        tp.read_properties_from_uri(self._thing_properties_uri)
        
        return self._datariver.create_thing(tp)
    
    def run(self):
        terminate = False
        print('Waiting for samples from ping to send back...')

        while not terminate:
            samples = self._thing.read_iot_nvp('Ping')

            for sample in samples:
                if sample.flow_state == FlowState.PURGED:
                    print('Received termination request. Terminating.')
                    terminate = True
                    break
                else:
                    self._thing.write('Pong', sample.data)


def main():
    try:
        pong = Pong("file://./config/PongProperties.json")
        # The 'with' statement supports a runtime context which is implemented
        # through a pair of methods executed (1) before the statement body is 
        # entered (__enter__()) and (2) after the statement body is exited (__exit__())
        with pong as p:
            p.run()
    except Exception as e:
        print('An unexpected error occurred: {}'.format(e))
    except KeyboardInterrupt:
        print('Ctrl-c detected .. exiting')
        sys.exit(1)

if __name__ == '__main__':
    main()





